src/tool/hpcrun/gpu/opencl/opencl-api.c

```diff
@@ -0,0 +1,460 @@
+// macros
+//******************************************************************************
+
+#define CPU_NANOTIME() (usec_time() * 1000)
+
+#define opencl_path() "libOpenCL.so"
+
+#define FORALL_OPENCL_ROUTINES(macro)	                                        \
+  macro(clGetEventProfilingInfo)   		                                        \
+  macro(clReleaseEvent)                                                       \
+  macro(clSetEventCallback)
+
+#define OPENCL_FN_NAME(f) DYN_FN_NAME(f)
+
+#define OPENCL_FN(fn, args)                                                   \
+  cl_int (*OPENCL_FN_NAME(fn)) args
+
+#define HPCRUN_OPENCL_CALL(fn, args)                                          \
+{                                                                             \
+  cl_int status = OPENCL_FN_NAME(fn) args;                                    \
+  if (status != CL_SUCCESS) {                                                 \
+    ETMSG(OPENCL, "opencl call failed: %s", opencl_error_report(status));     \
+  }                                                                           \
+}
+
+
+
+//----------------------------------------------------------
+// opencl function pointers for late binding
+//----------------------------------------------------------
+
+OPENCL_FN
+(
+  clGetEventProfilingInfo,
+  (
+    cl_event event,
+    cl_profiling_info param_name,
+    size_t param_value_size,
+    void* param_value,
+    size_t* param_value_size_ret
+  )
+);
+
+
+OPENCL_FN
+(
+  clReleaseEvent, 
+  (
+    cl_event event
+  )
+);
+
+
+OPENCL_FN
+(
+  clSetEventCallback,
+  (
+    cl_event event,
+    cl_int command_exec_callback_type,
+    void (CL_CALLBACK* pfn_notify)(cl_event event, cl_int event_command_status, void *user_data),
+    void* user_data
+  )
+);
+
+
+
+//******************************************************************************
+// local data
+//******************************************************************************
+
+static atomic_ullong opencl_pending_operations;
+
+
+
+//******************************************************************************
+// private operations
+//******************************************************************************
+
+static void
+opencl_pending_operations_adjust
+(
+  int value
+)
+{
+  atomic_fetch_add(&opencl_pending_operations, value);
+}
+
+
+static void
+opencl_activity_completion_notify
+(
+  void
+)
+{
+  gpu_monitoring_thread_activities_ready();
+}
+
+
+static void
+opencl_activity_process
+(
+  cl_event event,
+  void * user_data
+)
+{
+  gpu_activity_t gpu_activity;
+  opencl_activity_translate(&gpu_activity, event, user_data);
+  gpu_activity_process(&gpu_activity);
+}
+
+
+static void
+opencl_wait_for_pending_operations
+(
+  void
+)
+{
+  ETMSG(OPENCL, "pending operations: %lu", atomic_load(&opencl_pending_operations));
+  while (atomic_load(&opencl_pending_operations) != 0);
+}
+
+
+static const char*
```

Comment:
I don't like the fact that all of the OpenCL error code values are hardcoded into this file. Below is what I suggest instead. I thought that I posted this comment yesterday, but I don't see it or any change that would merit deleting it.

#define FORALL_OPENCL_ERRORS(macro)
  macro(CL_SUCCESS") \
  macro(CL_DEVICE_NOT_FOUND") \
  macro(CL_DEVICE_NOT_AVAILABLE") \
  macro(CL_COMPILER_NOT_AVAILABLE") \
  macro(CL_MEM_OBJECT_ALLOCATION_FAILURE") \
  macro(CL_OUT_OF_RESOURCES") \
  macro(CL_OUT_OF_HOST_MEMORY") \
  macro(CL_PROFILING_INFO_NOT_AVAILABLE") \
  macro(CL_MEM_COPY_OVERLAP") \
  macro(CL_IMAGE_FORMAT_MISMATCH") \
  macro(CL_IMAGE_FORMAT_NOT_SUPPORTED") \
  macro(CL_BUILD_PROGRAM_FAILURE") \
  macro(CL_MAP_FAILURE") \
  macro(CL_MISALIGNED_SUB_BUFFER_OFFSET") \
  macro(CL_EXEC_STATUS_ERROR_FOR_EVENTS_IN_WAIT_LIST") \
  macro(CL_COMPILE_PROGRAM_FAILURE") \
  macro(CL_LINKER_NOT_AVAILABLE") \
  macro(CL_LINK_PROGRAM_FAILURE") \
  macro(CL_DEVICE_PARTITION_FAILED") \
  macro(CL_KERNEL_ARG_INFO_NOT_AVAILABLE") \
  macro(CL_INVALID_VALUE") \
  macro(CL_INVALID_DEVICE_TYPE") \
  macro(CL_INVALID_PLATFORM") \
  macro(CL_INVALID_DEVICE") \
  macro(CL_INVALID_CONTEXT") \
  macro(CL_INVALID_QUEUE_PROPERTIES") \
  macro(CL_INVALID_COMMAND_QUEUE") \
  macro(CL_INVALID_HOST_PTR") \
  macro(CL_INVALID_MEM_OBJECT") \
  macro(CL_INVALID_IMAGE_FORMAT_DESCRIPTOR") \
  macro(CL_INVALID_IMAGE_SIZE") \
  macro(CL_INVALID_SAMPLER") \
  macro(CL_INVALID_BINARY") \
  macro(CL_INVALID_BUILD_OPTIONS") \
  macro(CL_INVALID_PROGRAM") \
  macro(CL_INVALID_PROGRAM_EXECUTABLE") \
  macro(CL_INVALID_KERNEL_NAME") \
  macro(CL_INVALID_KERNEL_DEFINITION") \
  macro(CL_INVALID_KERNEL") \
  macro(CL_INVALID_ARG_INDEX") \
  macro(CL_INVALID_ARG_VALUE") \
  macro(CL_INVALID_ARG_SIZE") \
  macro(CL_INVALID_KERNEL_ARGS") \
  macro(CL_INVALID_WORK_DIMENSION") \
  macro(CL_INVALID_WORK_GROUP_SIZE") \
  macro(CL_INVALID_WORK_ITEM_SIZE") \
  macro(CL_INVALID_GLOBAL_OFFSET") \
  macro(CL_INVALID_EVENT_WAIT_LIST") \
  macro(CL_INVALID_EVENT") \
  macro(CL_INVALID_OPERATION") \
  macro(CL_INVALID_GL_OBJECT") \
  macro(CL_INVALID_BUFFER_SIZE") \
  macro(CL_INVALID_MIP_LEVEL") \
  macro(CL_INVALID_GLOBAL_WORK_SIZE") \
  macro(CL_INVALID_PROPERTY") \
  macro(CL_INVALID_IMAGE_DESCRIPTOR") \
  macro(CL_INVALID_COMPILER_OPTIONS") \
  macro(CL_INVALID_LINKER_OPTIONS") \
  macro(CL_INVALID_DEVICE_PARTITION_COUNT") \
  macro(CL_INVALID_GL_SHAREGROUP_REFERENCE_KHR") \
  macro(CL_PLATFORM_NOT_FOUND_KHR") \
  macro(CL_INVALID_D3D10_DEVICE_KHR") \
  macro(CL_INVALID_D3D10_RESOURCE_KHR") \
  macro(CL_D3D10_RESOURCE_ALREADY_ACQUIRED_KHR") \
  macro(CL_D3D10_RESOURCE_NOT_ACQUIRED_KHR") 


#define FORALL_OPENCL_CALLS(macro) \
  macro(memcpy_H2D) \
  macro(memcpy_D2H) \
  macro(kernel)

#define CODE_TO_STRING(e) case e: return #e;



static const char*
opencl_call_to_string
(
  opencl_call_t type
)
{
  switch (type) {
  FORALL_OPENCL_CALLS(CODE_TO_STRING)
  default: return "CL_unknown_call";
  }
}


const char *
opencl_error_report
(
  cl_int error_status
)
{ 
  switch(error_status) {
  FORALL_OPENCL_ERRORS(CODE_TO_STRING)
  default: return "Unknown OpenCL error";
  }
}

Reply:
@jmellorcrummey . Got it. And I didnt see this comment yesterday, nor did I delete it

Reply:
I may have added it and forgot to press the "comment" button.

Reply:
How is this solution any better? We are just wrapping the error strings inside macros

Reply:
Notice that the numeric values of the codes and their binding with the error names are not embedded in the solution that I proposed.